src/main/java/se/liu/ida/hefquin/engine/queryproc/impl/poptimizer/simple/ReduceVocRewritingJoinPlanOptimizerImpl.java

```diff
@@ -0,0 +1,267 @@
+ * for which it uses a greedy approach to determine the join order based on vocabulary mapping and cardinality estimation.
+ * Specifically, choose the first subPlan with the smallest cardinality, then the subPlan with the same vocabulary mapping is selected.
+ * If there are multiple subPlans with the same vocabulary mapping, the order depends on estimated cardinality;
+ * If such a subPlan does not exist, simply select the subPlan with the lowest cardinality as the next subPlan.
+ *
+ * After determining join ordering, the physical plan can be constructed using symmetric hash join or bind join.
+ * {constructPlanWithSHJ} or {constructPlanWithBJ}
+ */
+public class ReduceVocRewritingJoinPlanOptimizerImpl extends JoinPlanOptimizerBase
+{
+    protected final FederationAccessManager fedAccessMgr;
+
+    public ReduceVocRewritingJoinPlanOptimizerImpl(final FederationAccessManager fedAccessMgr ) {
+        this.fedAccessMgr = fedAccessMgr;
+    }
+
+    @Override
+    public EnumerationAlgorithm initializeEnumerationAlgorithm( final List<PhysicalPlan> subplans ) {
+        return new GreedyConstructionAlgorithm(subplans);
+    }
+
+    protected class GreedyConstructionAlgorithm implements EnumerationAlgorithm {
+        protected final List<PhysicalPlan> subplans;
+
+        public GreedyConstructionAlgorithm( final List<PhysicalPlan> subplans ) {
+            this.subplans = subplans;
+        }
+
+        @Override
+        public PhysicalPlan getResultingPlan() throws PhysicalOptimizationException
+        {
+            // Extract all request operators from the given list of subplans,
+            // where the list of these operators is ordered in the order in which
+            // the operators can be found by a depth-first traversal of the subplans.
+            final List<LogicalOpRequest<?,?>> reqOpsOfAllSubPlans = new ArrayList<>();
+            for ( final PhysicalPlan subplan : subplans ) {
+            	reqOpsOfAllSubPlans.addAll(CardinalityBasedJoinPlanOptimizerUtils.extractAllRequestOpsFromSourceAssignment(subplan) );
+            }
+
+            // Next, get a list of cardinality estimates by performing cardinality
+            // requests for the above list of request operators.
+            final CardinalityResponse[] resps;
+            try {
+                resps = FederationAccessUtils.performCardinalityRequests(fedAccessMgr, reqOpsOfAllSubPlans);
+            } catch (final FederationAccessException e) {
+                throw new PhysicalOptimizationException("Issuing a cardinality request caused an exception.", e);
+            }
+
+            // Then, annotate each subplan (in the 'subplans') with all statistics relevant for the join planning
+            // ( including cardinality, number of access, and a list of relevant federation members )
+            final List<PhysicalPlanWithStatistics> subPlansWithStatistics = associateCardWithSubPlans(resps);
+
```

Comment:
This looks very familiar. Don't we have this code (and also the following lines) somewhere else already? If so, can you try to extract the common parts into a shared superclass method? Or some helper class?

Reply:
I have extracted the common parts into a superclass: 'CardinalityBasedJoinPlanOptimizerBase'